src/rfc3161_client/verify.py

```diff
@@ -12,170 +10,328 @@
 
-    return True
+        if self._intermediates:
+            verification_certificate.update(
+                cert.public_bytes(Encoding.DER) for cert in self._intermediates
+            )
 
+        p7 = tsp_response.time_stamp_token()
+        try:
+            self.verify_signed_data(p7, verification_certificate)
+        except ValueError as e:
+            msg = f"Error while verifying certificates: {e}"
+            raise VerificationError(msg)
 
-def verify_timestamp_response(
-    timestamp_response: TimeStampResponse, hashed_message: bytes, verify_opts: VerifyOpts
-) -> bool:
-    """Verify a Timestamp Response.
+        return True
 
-    Inspired by:
-        https://github.com/sigstore/timestamp-authority/blob/main/pkg/verification/verify.go#L209
+    def verify_signed_data(self, sig: bytes, certificates: set[bytes]) -> None:
+        """Verify signed data.
 
-    """
-    # Note: digitorus/timestamp does not validate if the result is GRANTED_WITH_MOD
-    # https://github.com/digitorus/timestamp/blob/master/timestamp.go#L268
-    if PKIStatus(timestamp_response.status) != PKIStatus.GRANTED:
-        msg = "PKIStatus is not GRANTED"
-        raise VerificationError(msg)
+        This function verifies that the bytes used in a signature are signed by a certificate
+        trusted in the `certificates` list.
+        The function does not return anything, but raises an exception if the verification fails.
 
-    _verify_tsr_with_chains(timestamp_response, verify_opts)
+        :param sig: Bytes of a PKCS7 object. This must be in DER format and will be unserialized.
+        :param certificates: A list of trusted certificates to verify the response against.
+        :raise: ValueError if the signature verification fails.
+        """
+        return _rust_verify.pkcs7_verify(sig, list(certificates))
 
-    # Verify Nonce
-    if verify_opts.nonce is not None and timestamp_response.tst_info.nonce != verify_opts.nonce:
-        msg = "Nonce mismatch"
-        raise VerificationError(msg)
 
-    # Verify Policy ID
-    if (
-        verify_opts.policy_id is not None
-        and timestamp_response.tst_info.policy != verify_opts.policy_id
-    ):
-        msg = "Policy ID mismatch"
-        raise VerificationError(msg)
+def create_verifier_from_request(
+    tsp_request: TimeStampRequest,
+    tsa_certificate: cryptography.x509.Certificate | None,
+    common_name: str | None,
+    root_certificates: list[cryptography.x509.Certificate] | None = None,
+    intermediates: list[cryptography.x509.Certificate] | None = None,
+) -> Verifier:
+    if intermediates is None:
+        intermediates = []
+
+    if root_certificates is None:
+        root_certificates = []
+
+    builder = VerifyBuilder()
+    if tsp_request.policy:
+        builder = builder.policy_id(tsp_request.policy)
+    if tsp_request.nonce:
+        builder = builder.nonce(tsp_request.nonce)
```

Comment:
Are these checks necessary? can `TimestampRequest.policy` and `TimestampRequest.nonce` be None?

Reply:
the request fields are optional?

Reply:
I'm asking because of these type signatures: https://github.com/trailofbits/rfc3161-client/blob/734b1228396bf00d984ed8a1f806125d6e37054d/src/rfc3161_client/tsp.py#L40-L48

Reply:
Good catch - the signature here should be updated!